### PR TITLE
Generalize polynomial type

### PR DESCRIPTION
**Pull request by ryanleh**
Polynomial generalization w/o `marlin_pst13` scheme

### CODE REVIEW

File: src/ipa_pc/mod.rs [hunk 1]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
-
+    use ark_ec::AffineCurve;
     use ark_ed_on_bls12_381::EdwardsAffine;
+    use ark_ff::UniformRand;
+    use ark_poly::{univariate::DensePolynomial as DensePoly, UVPolynomial};
     use blake2::Blake2s;
 
-    type PC<E, D> = InnerProductArgPC<E, D>;
-    type PC_JJB2S = PC<EdwardsAffine, Blake2s>;
+    type UniPoly = DensePoly<<EdwardsAffine as AffineCurve>::ScalarField>;
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1043]:
```suggestion
    type UniPoly = DensePoly<ark_ed_on_bls12_381::Fr>;
```

File: src/ipa_pc/mod.rs [hunk 2]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
+    ) -> DensePoly<A::ScalarField> {
+        DensePoly::<A::ScalarField>::rand(degree, rng)
+    }
+
+    fn rand_point<A: AffineCurve>(
+        _: Option<usize>,
+        rng: &mut rand::prelude::StdRng,
+    ) -> A::ScalarField {
+        A::ScalarField::rand(rng)
+    }
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1060]:
```suggestion
    fn rand_point<F: Field>(
        _: Option<usize>,
        rng: &mut rand::prelude::StdRng,
    ) -> F {
        F::rand(rng)
    }
```

File: src/ipa_pc/mod.rs [hunk 3]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
+    type PC<E, D, P> = InnerProductArgPC<E, D, P>;
+    type PC_JJB2S = PC<EdwardsAffine, Blake2s, UniPoly>;
+
+    fn rand_poly<A: AffineCurve>(
+        degree: usize,
+        _: Option<usize>,
+        rng: &mut rand::prelude::StdRng,
+    ) -> DensePoly<A::ScalarField> {
+        DensePoly::<A::ScalarField>::rand(degree, rng)
+    }
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1053]:
```suggestion
    fn rand_poly<F: PrimeField>(
        degree: usize,
        _: Option<usize>,
        rng: &mut rand::prelude::StdRng,
    ) -> DensePoly<F> {
        DensePoly::rand(degree, rng)
    }
```

File: src/ipa_pc/mod.rs [hunk 4]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
 
     #[test]
     fn single_poly_test() {
         use crate::tests::*;
-        single_poly_test::<_, PC_JJB2S>().expect("test failed for ed_on_bls12_381-blake2s");
+        single_poly_test::<_, _, PC_JJB2S>(
+            None,
+            rand_poly::<EdwardsAffine>,
+            rand_point::<EdwardsAffine>,
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1068]:
```suggestion
            rand_poly::<Fr>,
            rand_point::<Fr>,
```

File: src/ipa_pc/mod.rs [hunk 5]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
     }
 
     #[test]
     fn quadratic_poly_degree_bound_multiple_queries_test() {
         use crate::tests::*;
-        quadratic_poly_degree_bound_multiple_queries_test::<_, PC_JJB2S>()
-            .expect("test failed for ed_on_bls12_381-blake2s");
+        quadratic_poly_degree_bound_multiple_queries_test::<_, _, PC_JJB2S>(
+            rand_poly::<EdwardsAffine>,
+            rand_point::<EdwardsAffine>,
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1078]:
```suggestion
            rand_poly::<Fr>,
            rand_point::<Fr>,
```

File: src/ipa_pc/mod.rs [hunk 6]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
     }
 
     #[test]
     fn linear_poly_degree_bound_test() {
         use crate::tests::*;
-        linear_poly_degree_bound_test::<_, PC_JJB2S>()
-            .expect("test failed for ed_on_bls12_381-blake2s");
+        linear_poly_degree_bound_test::<_, _, PC_JJB2S>(
+            rand_poly::<EdwardsAffine>,
+            rand_point::<EdwardsAffine>,
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1088]:
```suggestion
            rand_poly::<Fr>,
            rand_point::<Fr>,
```

File: src/ipa_pc/mod.rs [hunk 7]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
     }
 
     #[test]
     fn single_poly_degree_bound_test() {
         use crate::tests::*;
-        single_poly_degree_bound_test::<_, PC_JJB2S>()
-            .expect("test failed for ed_on_bls12_381-blake2s");
+        single_poly_degree_bound_test::<_, _, PC_JJB2S>(
+            rand_poly::<EdwardsAffine>,
+            rand_point::<EdwardsAffine>,
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1098]:
```suggestion
            rand_poly::<Fr>,
            rand_point::<Fr>,
```

File: src/ipa_pc/mod.rs [hunk 8]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
     }
 
     #[test]
     fn single_poly_degree_bound_multiple_queries_test() {
         use crate::tests::*;
-        single_poly_degree_bound_multiple_queries_test::<_, PC_JJB2S>()
-            .expect("test failed for ed_on_bls12_381-blake2s");
+        single_poly_degree_bound_multiple_queries_test::<_, _, PC_JJB2S>(
+            rand_poly::<EdwardsAffine>,
+            rand_point::<EdwardsAffine>,
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1108]:
```suggestion
            rand_poly::<Fr>,
            rand_point::<Fr>,
```

File: src/ipa_pc/mod.rs [hunk 9]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
...
     }
 
     #[test]
     fn two_polys_degree_bound_single_query_test() {
         use crate::tests::*;
-        two_polys_degree_bound_single_query_test::<_, PC_JJB2S>()
-            .expect("test failed for ed_on_bls12_381-blake2s");
+        two_polys_degree_bound_single_query_test::<_, _, PC_JJB2S>(
+            rand_poly::<EdwardsAffine>,
+            rand_point::<EdwardsAffine>,
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1118]:
```suggestion
            rand_poly::<Fr>,
            rand_point::<Fr>,
```

File: src/ipa_pc/mod.rs [hunk 10]
```diff
@@ -1028,96 +1034,160 @@ mod tests {
     #![allow(non_camel_case_types)]
 
     use super::InnerProductArgPC;
-
+    use ark_ec::AffineCurve;
     use ark_ed_on_bls12_381::EdwardsAffine;
```
Comment (Pratyush) [inline src/ipa_pc/mod.rs:1038]:
```suggestion
    use ark_ed_on_bls12_381::{EdwardsAffine, Fr};
```

File: src/kzg10/data_structures.rs [hunk 11]
```diff
@@ -209,12 +212,13 @@ impl<E: PairingEngine> PreparedCommitment<E> {
     PartialEq(bound = ""),
     Eq(bound = "")
 )]
-pub struct Randomness<E: PairingEngine> {
+pub struct Randomness<E: PairingEngine, P: UVPolynomial<E::Fr>> {
```
Comment (Pratyush) [inline src/kzg10/data_structures.rs:215]:
```suggestion
pub struct Randomness<F: PrimeField, P: UVPolynomial<F>> {
```

File: src/kzg10/mod.rs [hunk 12]
```diff
@@ -536,23 +560,28 @@ mod tests {
...
             let mut degree = 0;
             while degree <= 1 {
                 degree = usize::rand(rng) % 20;
             }
-            let pp = KZG10::<E>::setup(degree, false, rng)?;
-            let (ck, vk) = KZG10::trim(&pp, degree)?;
-            let p = Polynomial::rand(degree, rng);
+            let pp = KZG10::<E, P>::setup(degree, false, rng)?;
+            let (ck, vk) = KZG10::<E, P>::trim(&pp, degree)?;
+            let p = rand_poly::<E, P>(degree, None, rng);
```
Comment (Pratyush) [inline src/kzg10/mod.rs:577]: Why does `Polynomial::rand` no longer work?

Comment (ryanleh) [reply src/kzg10/mod.rs:577]: Oh I guess for kzg templates this would be fine good point.

File: src/lib.rs [hunk 13]
```diff
@@ -879,25 +903,34 @@ pub mod tests {
         Ok(())
     }
 
-    fn equation_test_template<F, PC>(info: TestInfo) -> Result<(), PC::Error>
+    fn equation_test_template<F, P, PC>(
+        info: TestInfo,
+        rand_poly: fn(usize, Option<usize>, &mut rand::prelude::StdRng) -> P,
+        rand_point: fn(Option<usize>, &mut rand::prelude::StdRng) -> P::Point,
```
Comment (Pratyush) [inline src/lib.rs:909]: Let's roll `rand_poly` and `rand_point` into `TestInfo`?

File: src/sonic_pc/data_structures.rs [hunk 14]
```diff
@@ -8,7 +8,7 @@ use ark_ec::{PairingEngine, ProjectiveCurve};
 pub type UniversalParams<E> = kzg10::UniversalParams<E>;
 
 /// `Randomness` is the randomness for the KZG10 scheme.
-pub type Randomness<E> = kzg10::Randomness<E>;
+pub type Randomness<E, P> = kzg10::Randomness<E, P>;
```
Comment (Pratyush) [inline src/sonic_pc/data_structures.rs:11]: Do we want to give the same treatment here, or should we reserve that for a future PR?

File: Cargo.toml [hunk 15]
```diff
@@ -42,6 +42,12 @@ ark-bls12-381 = { git = "https://github.com/arkworks-rs/curves", default-feature
 ark-bls12-377 = { git = "https://github.com/arkworks-rs/curves", default-features = false, features = [ "curve" ] }
 blake2 = { version = "0.8", default-features = false }
 
+[patch.'https://github.com/arkworks-rs/algebra']
+ark-serialize = { git = "https://github.com/ryanleh/algebra", branch = "multivariate-support", default-features = false }
+ark-ff = { git = "https://github.com/ryanleh/algebra", branch = "multivariate-support", default-features = false }
+ark-ec = { git = "https://github.com/ryanleh/algebra", branch = "multivariate-support", default-features = false }
+ark-poly = { git = "https://github.com/ryanleh/algebra", branch = "multivariate-support", default-features = false }
+
```
Comment (Pratyush) [inline Cargo.toml:50]: TODO: remove this once respective PR is merged

```suggestion
```

File: src/lib.rs [hunk 16]
```diff
@@ -84,19 +85,19 @@ pub mod ipa_pc;
...
 /// that `p[label]` is to be queried at.
-pub type QuerySet<'a, F> = BTreeSet<(String, (String, F))>;
+pub type QuerySet<'a, T> = BTreeSet<(String, (String, T))>;
 
 /// `Evaluations` is the result of querying a set of labeled polynomials or equations
 /// `p` at a `QuerySet` `Q`. It maps each element of `Q` to the resulting evaluation.
 /// That is, if `(label, query)` is an element of `Q`, then `evaluation.get((label, query))`
 /// should equal `p[label].evaluate(query)`.
-pub type Evaluations<'a, F> = BTreeMap<(String, F), F>;
+pub type Evaluations<'a, F, T> = BTreeMap<(String, T), F>;
```
Comment (Pratyush) [inline src/lib.rs:96]: Let's make `T` come first, since it's the key to the map.

```suggestion
pub type Evaluations<'T, F> = BTreeMap<(String, T), F>;
```

File: src/lib.rs [hunk 17]
```diff
@@ -84,19 +85,19 @@ pub mod ipa_pc;
 /// `QuerySet` is the set of queries that are to be made to a set of labeled polynomials/equations
 /// `p` that have previously been committed to. Each element of a `QuerySet` is a pair of
 /// `(label, (point_label, point))`, where `label` is the label of a polynomial in `p`,
-/// `point_label` is the label for the point (e.g., "beta"), and  and `point` is the field element
+/// `point_label` is the label for the point (e.g., "beta"), and  and `point` is the location
 /// that `p[label]` is to be queried at.
-pub type QuerySet<'a, F> = BTreeSet<(String, (String, F))>;
+pub type QuerySet<'a, T> = BTreeSet<(String, (String, T))>;
```
Comment (Pratyush) [inline src/lib.rs:90]: Drive-by clean up

```suggestion
pub type QuerySet<T> = BTreeSet<(String, (String, T))>;
```
